File: internal/keystore/keystore.go
// Package keystore provides filesystem storage for AirPods encryption keys.
//
// Keys are stored in ~/.local/share/linuxpods/keys.json following the
// XDG Base Directory specification.
//
// Storage format:
//   - File format: JSON
//   - Data structure: {"version": 1, "keys": {"MAC": "base64-key", ...}}
//
// Usage:
//
//	store, err := keystore.New()
//	if err != nil {
//		log.Fatal(err)
//	}
//
//	// Save a key
//	key := []byte{0x01, 0x02, ...} // 16-byte encryption key
//	err = store.Set("AA:BB:CC:DD:EE:FF", key)
//	err = store.Save()
//
//	// Load all keys
//	keys, err := store.Load()
//
//	// Get a specific key
//	key, exists := store.Get("AA:BB:CC:DD:EE:FF")
package keystore

import (
	"encoding/base64"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
)

const (
	// XDG_DATA_HOME default path
	xdgDataHome = ".local/share"
	appName     = "linuxpods"
	keysFile    = "keys.json"
)

// Keystore manages storage of AirPods encryption keys
type Keystore struct {
	dataDir string            // XDG data directory path
	keys    map[string][]byte // In-memory cache: MAC address -> encryption key
}

// keyData represents the JSON structure of the stored keys
type keyData struct {
	Version int               `json:"version"`
	Keys    map[string]string `json:"keys"` // MAC address -> base64-encoded key
}

// New creates a new Keystore instance
// It creates the data directory if it doesn't exist
func New() (*Keystore, error) {
	// Determine XDG data directory
	dataDir, err := getDataDir()
	if err != nil {
		return nil, fmt.Errorf("failed to determine data directory: %w", err)
	}

	// Create the directory if it doesn't exist
	if err := os.MkdirAll(dataDir, 0700); err != nil {
		return nil, fmt.Errorf("failed to create data directory: %w", err)
	}

	ks := &Keystore{
		dataDir: dataDir,
		keys:    make(map[string][]byte),
	}

	return ks, nil
}

// Load reads all keys from disk
// Returns a map of MAC address -> encryption key
// If the file doesn't exist, returns an empty map (not an error)
func (ks *Keystore) Load() (map[string][]byte, error) {
	filePath := filepath.Join(ks.dataDir, keysFile)

	if _, err := os.Stat(filePath); os.IsNotExist(err) {
		// File doesn't exist - return an empty map
		return make(map[string][]byte), nil
	}

	data, err := os.ReadFile(filePath)
	if err != nil {
		return nil, fmt.Errorf("failed to read keys file: %w", err)
	}

	var keyData keyData
	if err := json.Unmarshal(data, &keyData); err != nil {
		return nil, fmt.Errorf("failed to parse keys JSON: %w", err)
	}

	// Decode base64 keys
	keys := make(map[string][]byte)
	for mac, keyB64 := range keyData.Keys {
		key, err := base64.StdEncoding.DecodeString(keyB64)
		if err != nil {
			return nil, fmt.Errorf("failed to decode key for %s: %w", mac, err)
		}
		keys[mac] = key
	}

	ks.keys = keys
	return keys, nil
}

// Save writes all keys to disk
func (ks *Keystore) Save() error {
	keysB64 := make(map[string]string)
	for mac, key := range ks.keys {
		keysB64[mac] = base64.StdEncoding.EncodeToString(key)
	}

	data := keyData{
		Version: 1,
		Keys:    keysB64,
	}

	jsonData, err := json.MarshalIndent(data, "", "  ")
	if err != nil {
		return fmt.Errorf("failed to marshal keys to JSON: %w", err)
	}

	// Write (owner read/write-only)
	filePath := filepath.Join(ks.dataDir, keysFile)
	if err := os.WriteFile(filePath, jsonData, 0600); err != nil {
		return fmt.Errorf("failed to write keys file: %w", err)
	}

	return nil
}

// Get retrieves a key for a specific MAC address from the in-memory cache
// Returns the key and true if found, or nil and false if not found
// Call Load() first to populate the cache from disk
func (ks *Keystore) Get(macAddr string) ([]byte, bool) {
	key, exists := ks.keys[macAddr]
	if !exists {
		return nil, false
	}

	// Return a copy to prevent external modification
	keyCopy := make([]byte, len(key))
	copy(keyCopy, key)
	return keyCopy, true
}

// Set stores a key for a specific MAC address in the in-memory cache
// Call Save() to persist to disk
func (ks *Keystore) Set(macAddr string, key []byte) error {
	if len(key) == 0 {
		return fmt.Errorf("encryption key cannot be empty")
	}

	// Store a copy to prevent external modification
	keyCopy := make([]byte, len(key))
	copy(keyCopy, key)
	ks.keys[macAddr] = keyCopy

	return nil
}

// Delete removes a key for a specific MAC address from the in-memory cache
// Call Save() to persist to disk
func (ks *Keystore) Delete(macAddr string) {
	delete(ks.keys, macAddr)
}

// List returns all MAC addresses that have stored keys
func (ks *Keystore) List() []string {
	macs := make([]string, 0, len(ks.keys))
	for mac := range ks.keys {
		macs = append(macs, mac)
	}
	return macs
}

// Clear removes all keys from the in-memory cache
// Call Save() to persist to disk
func (ks *Keystore) Clear() {
	ks.keys = make(map[string][]byte)
}

// GetAll returns a copy of all stored keys
func (ks *Keystore) GetAll() map[string][]byte {
	keys := make(map[string][]byte, len(ks.keys))
	for mac, key := range ks.keys {
		keyCopy := make([]byte, len(key))
		copy(keyCopy, key)
		keys[mac] = keyCopy
	}
	return keys
}

// getDataDir returns the XDG data directory for the application
// Follows XDG Base Directory specification:
//   - Uses $XDG_DATA_HOME if set
//   - Falls back to ~/.local/share
func getDataDir() (string, error) {
	// Check XDG_DATA_HOME environment variable
	dataHome := os.Getenv("XDG_DATA_HOME")
	if dataHome == "" {
		// Use default: ~/.local/share
		home, err := os.UserHomeDir()
		if err != nil {
			return "", fmt.Errorf("failed to get home directory: %w", err)
		}
		dataHome = filepath.Join(home, xdgDataHome)
	}

	appDataDir := filepath.Join(dataHome, appName)
	return appDataDir, nil
}

File: src/keystore.rs
//! Persistent storage for AirPods BLE encryption keys.
//!
//! Stored at ~/.local/share/linuxpods/keys.json (XDG Base Directory), mode 0600:
//! {"version": 1, "keys": {"MAC": "base64-key", ...}}

use std::collections::HashMap;
use std::path::{Path, PathBuf};

use anyhow::{Context, Result};
use base64::Engine as _;
use base64::engine::general_purpose::STANDARD as B64;
use serde::{Deserialize, Serialize};

const APP_NAME: &str = "linuxpods";
const KEYS_FILE: &str = "keys.json";
const FORMAT_VERSION: u32 = 1;

#[derive(Serialize, Deserialize)]
struct KeyData {
    version: u32,
    /// MAC address -> base64-encoded key
    keys: HashMap<String, String>,
}

/// In-memory cache plus explicit load/save, mirroring the Go API.
pub struct Keystore {
    data_dir: PathBuf,
    keys: HashMap<String, Vec<u8>>,
}

impl Keystore {
    /// Creates the data directory (0700) if absent.
    pub fn new() -> Result<Self> {
        let data_dir = data_dir()?;
        std::fs::create_dir_all(&data_dir)
            .with_context(|| format!("failed to create data directory {}", data_dir.display()))?;
        restrict_permissions(&data_dir, 0o700)?;
        Ok(Self {
            data_dir,
            keys: HashMap::new(),
        })
    }

    /// Test hook: use an explicit directory instead of the XDG one.
    pub fn with_dir(data_dir: PathBuf) -> Result<Self> {
        std::fs::create_dir_all(&data_dir)?;
        Ok(Self {
            data_dir,
            keys: HashMap::new(),
        })
    }

    fn path(&self) -> PathBuf {
        self.data_dir.join(KEYS_FILE)
    }

    /// Reads all keys from disk. A missing file is not an error.
    pub fn load(&mut self) -> Result<HashMap<String, Vec<u8>>> {
        let path = self.path();
        if !path.exists() {
            return Ok(HashMap::new());
        }

        let raw = std::fs::read(&path)
            .with_context(|| format!("failed to read keys file {}", path.display()))?;
        let parsed: KeyData = serde_json::from_slice(&raw).context("failed to parse keys JSON")?;

        let mut keys = HashMap::with_capacity(parsed.keys.len());
        for (mac, b64) in parsed.keys {
            let key = B64
                .decode(b64.as_bytes())
                .with_context(|| format!("failed to decode key for {mac}"))?;
            keys.insert(mac, key);
        }

        self.keys = keys.clone();
        Ok(keys)
    }

    /// Writes all cached keys to disk with 0600 permissions.
    pub fn save(&self) -> Result<()> {
        let data = KeyData {
            version: FORMAT_VERSION,
            keys: self
                .keys
                .iter()
                .map(|(mac, key)| (mac.clone(), B64.encode(key)))
                .collect(),
        };

        let json = serde_json::to_vec_pretty(&data).context("failed to marshal keys to JSON")?;
        let path = self.path();
        std::fs::write(&path, json)
            .with_context(|| format!("failed to write keys file {}", path.display()))?;
        restrict_permissions(&path, 0o600)?;
        Ok(())
    }

    pub fn get(&self, mac: &str) -> Option<Vec<u8>> {
        self.keys.get(mac).cloned()
    }

    /// Rejects empty keys, as the Go version did.
    pub fn set(&mut self, mac: &str, key: &[u8]) -> Result<()> {
        anyhow::ensure!(!key.is_empty(), "encryption key cannot be empty");
        self.keys.insert(mac.to_string(), key.to_vec());
        Ok(())
    }

    pub fn remove(&mut self, mac: &str) {
        self.keys.remove(mac);
    }

    pub fn list(&self) -> Vec<String> {
        self.keys.keys().cloned().collect()
    }

    pub fn clear(&mut self) {
        self.keys.clear();
    }

    pub fn all(&self) -> HashMap<String, Vec<u8>> {
        self.keys.clone()
    }
}

/// $XDG_DATA_HOME/linuxpods, falling back to ~/.local/share/linuxpods.
fn data_dir() -> Result<PathBuf> {
    let base = dirs::data_dir().context("failed to determine XDG data directory")?;
    Ok(base.join(APP_NAME))
}

#[cfg(unix)]
fn restrict_permissions(path: &Path, mode: u32) -> Result<()> {
    use std::os::unix::fs::PermissionsExt;
    let perms = std::fs::Permissions::from_mode(mode);
    std::fs::set_permissions(path, perms)
        .with_context(|| format!("failed to set permissions on {}", path.display()))
}

#[cfg(not(unix))]
fn restrict_permissions(_path: &Path, _mode: u32) -> Result<()> {
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    fn store() -> (Keystore, tempfile::TempDir) {
        let dir = tempfile::tempdir().unwrap();
        (Keystore::with_dir(dir.path().to_path_buf()).unwrap(), dir)
    }

    #[test]
    fn missing_file_loads_empty() {
        let (mut ks, _d) = store();
        assert!(ks.load().unwrap().is_empty());
    }

    #[test]
    fn round_trips_through_disk() {
        let (mut ks, dir) = store();
        let key = vec![1u8; 16];
        ks.set("AA:BB:CC:DD:EE:FF", &key).unwrap();
        ks.save().unwrap();

        let mut reopened = Keystore::with_dir(dir.path().to_path_buf()).unwrap();
        let loaded = reopened.load().unwrap();
        assert_eq!(loaded.get("AA:BB:CC:DD:EE:FF"), Some(&key));
    }

    #[test]
    fn rejects_empty_key() {
        let (mut ks, _d) = store();
        assert!(ks.set("AA:BB:CC:DD:EE:FF", &[]).is_err());
    }

    #[test]
    fn reads_go_written_format() {
        let dir = tempfile::tempdir().unwrap();
        std::fs::write(
            dir.path().join(KEYS_FILE),
            r#"{"version":1,"keys":{"AA:BB:CC:DD:EE:FF":"AQIDBAUGBwgJCgsMDQ4PEA=="}}"#,
        )
        .unwrap();

        let mut ks = Keystore::with_dir(dir.path().to_path_buf()).unwrap();
        let loaded = ks.load().unwrap();
        assert_eq!(loaded["AA:BB:CC:DD:EE:FF"], (1u8..=16).collect::<Vec<u8>>());
    }

    #[test]
    fn save_uses_owner_only_permissions() {
        use std::os::unix::fs::PermissionsExt;
        let (mut ks, dir) = store();
        ks.set("AA:BB:CC:DD:EE:FF", &[9u8; 16]).unwrap();
        ks.save().unwrap();
        let mode = std::fs::metadata(dir.path().join(KEYS_FILE))
            .unwrap()
            .permissions()
            .mode();
        assert_eq!(mode & 0o777, 0o600);
    }
}
